// jd3t2/publish/SanitizeDoclets.js
var SanitizeDoclets = function(publisher){
		this.publisher = publisher;
		this.db = this.publisher.db;
	},
	path = require('path')
;

SanitizeDoclets.prototype = {
	publisher: null,
	db: null,

	process: function(){
		var count = this.db().count(),
			me = this
		;
		this.publisher.log('info', 'Initializing Sanitization');
		this.publisher.log('debug', 'Doclets to parse: ' + count);

		var step = Math.ceil(count / 10);
		if(step > 50){step = 50;}

		this.db().each(function(doclet, ix){
			if((ix%step) === 0){
				me.publisher.log('debug', 'Iterated: ' + ix + '/' + count);
			}
			try{
				me.processDoclet(doclet);
			}catch(e){
				me.publisher.log('error', 'Processing doclet: ' + doclet.longname, e);
				me.publisher.log('error', e);
			}
		});
	},

	processDoclet: function(doclet){
		if(undefined === doclet.memberof){
			doclet.memberof = '__global__'; //create de fucking key to filter with TAFFYDB
		}
		if((doclet.kind === 'function' || doclet.kind === 'event') && !doclet.returns){
			doclet.returns = [{
				type: {
					names: ['void'],
					optional: [],
					nullable: [],
					variable: []
				}
			}];
		}

		if(doclet.kind === 'file'){
			this.publisher.Files.addDocletToFile(doclet);
		}
		// process File path
		if(doclet.meta && doclet.meta.filename){
			if(undefined === doclet.meta.path){
				doclet.meta.real_file = path.normalize( doclet.meta.path + '/' + doclet.meta.filename );
				doclet.meta.path = this.publisher.Files.addFile(doclet.meta.real_file);
			}
		}

//@todo		this.processDocletTags(doclet);
//@todo		this.processDocletProperties(doclet);

		

	}
};

module.exports = SanitizeDoclets;
