// jsdocGUI/GUI/Actions.js
/**
 * @class Actions
 * @classdesc Handle abstract action and asociate this actions to DOM elements
 * @desc Asociate the actions registred via addAction to DOM elements using 
 * class ```.[ACTION_NAME_IN_UPPER_CASE]```
 */
function Actions(jQuery){
	this.$ = jQuery;
	this.actionNames = [];
	this.actions = {};
}

/**
 * @private
 * @desc Transform an action to ACTION
 * @param {String} action Action to been sanitized
 * @return {String}
 */
function sanitize(action){
	return action.toUpperCase().replace(/[\s\W]/g, '_');
}

/**
 * @method
 * @desc Add an action to the list actions
 * 
 * @param {String} action Name of the action (transformed to upperCase)
 * @param {ActionCallback} cbk Callback associated to the action
 */
Actions.prototype.addAction = function(action, cbk){
	action = sanitize(action);
	this.actionNames.push(action);
	this.actions[action] = cbk;
	
	this.associate(action);
}

/**
 * @public
 * @method
 * @desc Execute the action passed, if not $event passed, creates one fakeEvent
 * 
 * @param {String} action Name of the action to be executed
 * @param {jQuery.Event} $event Event to be passed to callback;
 */
Actions.prototype.exec = function(action, $event){
	action = sanitize(action);
	if('function' === typeof(this.actions[action])){
		
		if(undefined === $event){
			$event = new this.$.Event('fakeEvent');
		}
		
		return this.actions[action].call(
			this.$('.' + action), 
			$event
		);
	}
}

/**
 * @protected
 * @desc Associate the action to the DOM elements
 * 
 * @param {String} action Name of the action to be associated
 */
Actions.prototype.associate = function(action){
	var actionObj = this;
	this.$('.' + action).live('click', (function(act){
		return function($event){
			return actionObj.exec(act, $event);
		};
	}(action)));
};

module.exports = Actions;

/**
 * @name ActionCallback
 * @function
 * @param {jQuery.Event} event jQuery Event, if the action is executed via user 
 *         event, you get a click event, if is an action executed via code, you 
 *         get a fakeEvent
 * @this {jQuery} All elements with class ACTION
 */
