// jsdocGUI/GUI/jsdocGUI.js
(function(){
	var nwWindow = nwrequire('nw.gui').Window.get(),
		$ = nwWindow.window.jQuery,
		document = nwWindow.window.document,
		Actions = require('./Actions'),
		jsdocGUI
	;
	//@todo Document!
	jsdocGUI = {
		actions: new Actions($),
		modals: {},
		
		init: function(){
			this.addEventListeners();
			this.createActions();
			this.bindKeys();
			this.createModals();
			
			
			//@todo REMOVE FROM HERE, create a compononet for this
			console.log('Adding functionality to browse directory buttons');
			$('input[type="file"]').each(function(){
				var chooser = $(this);
				chooser.attr('nwworkingdir', process.cwd());
				chooser.siblings('button').click(function(){
					
					chooser.trigger('click');            
					chooser.change(function(evt) {
						if(chooser.val() != ''){
							chooser.attr('nwworkingdir', chooser.val());
							chooser.siblings('input').val(chooser.val());
							console.log($(this).val());
						}
					});
				});
			});
			
		},
		
		_reset: function(){
			var module;
			for(module in require.cache){
				if(require.cache.hasOwnProperty(module)){
					delete require.cache[module];
				}
			}
			nwWindow.reloadIgnoringCache();
		},
		
		addEventListeners: function(){
			console.log('Preventing Close with out confirmation');
			nwWindow.on('close', function() {
				jsdocGUI.modals.close.modal('show');
				return false;
			});
		},
		createActions: function(){
			console.log('Creating Actions');
			this.actions.addAction('CLOSE_ACTION', function(){
				nwWindow.close();
			});
			this.actions.addAction('FORCE_CLOSE_ACTION', function(){
				nwWindow.close(true);
			});
			this.actions.addAction('MINIMIZE_ACTION', function(){
				nwWindow.minimize();
			});
			
			if(global.DEBUG_MODE){
				this.actions.addAction('DEVTOOLS_ACTION', function(){
					nwWindow.showDevTools();
				});
				this.actions.addAction('REFRESH_ACTION', this._reset);
				$('.REFRESH_ACTION,.DEVTOOLS_ACTION').removeClass('hidden');
			}
		},
		
		bindKeys: function(){
			console.log('Binding Keys');
			$(document).bind('Key.F12', function(event){
				jsdocGUI.actions.exec('DEVTOOLS_ACTION');
			});
			$(document).bind('Key.Esc', function(event){
				jsdocGUI.actions.exec('CLOSE_ACTION');
			});
			$(document).bind('Key.F5', function(event){
				jsdocGUI.actions.exec('REFRESH_ACTION');
			});
		},
		
		createModals: function(){
			console.log('Creating Modals');
			this.modals.close = $('#closeModal').modal({
				backdrop: true,
				keyboard: false,
				show: false
			});
		}
	};
	
	module.exports = jsdocGUI;
}());
